ambry-commons/src/main/java/com.github.ambry.commons/RetainingAsyncWritableChannel.java

```diff
@@ -0,0 +1,178 @@
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+
+/**
+ * An implementation of {@link AsyncWritableChannel} that will retain a reference to any {@link ByteBuf}s passed to it.
+ * It will assemble these buffers into a {@link CompositeByteBuf} that can then be consumed by the user of the channel.
+ * Meant for interacting with blocking content consumers that would otherwise require a background thread if they did
+ * not have all of the content available.
+ */
+public class RetainingAsyncWritableChannel implements AsyncWritableChannel {
+  private final long sizeLimitInBytes;
+  private volatile boolean open = true;
+  private final AtomicLong totalBytesWritten = new AtomicLong(0);
+  // accesses/updates to the buffer under construction must be protected by the lock
+  private final Object bufferLock = new Object();
+  private CompositeByteBuf compositeBuffer = ByteBufAllocator.DEFAULT.compositeBuffer();
+
+  /**
+   * Construct a {@link RetainingAsyncWritableChannel} with the size limit set to {@link Long#MAX_VALUE}.
+   */
+  public RetainingAsyncWritableChannel() {
+    this(Integer.MAX_VALUE);
+  }
+
+  /**
+   * @param sizeLimitInBytes the maximum number of bytes that can be written to this channel. If this limit is exceeded,
+   *                         a {@link RestServiceException} will be provided to the write callback.
+   */
+  public RetainingAsyncWritableChannel(int sizeLimitInBytes) {
+    this.sizeLimitInBytes = sizeLimitInBytes;
+  }
+
+  /**
+   * @inheritDoc
+   *
+   * This method requires copying
+   */
+  @Override
+  public Future<Long> write(ByteBuffer src, Callback<Long> callback) {
+    // still need to copy the buffer since the writer may decide to reuse the buffer after the callback is called.
+    return writeInternal(() -> {
+      ByteBuf copy = src.isDirect() ? ByteBufAllocator.DEFAULT.directBuffer(src.remaining())
+          : PooledByteBufAllocator.DEFAULT.heapBuffer(src.remaining());
+      copy.writeBytes(src);
+      return copy;
+    }, callback);
+  }
+
+  @Override
+  public Future<Long> write(ByteBuf src, Callback<Long> callback) {
+    return writeInternal(() -> {
+      ByteBuf duplicate = src.retainedDuplicate();
+      // mark all bytes as read.
+      src.skipBytes(src.readableBytes());
+      return duplicate;
+    }, callback);
+  }
+
+  private Future<Long> writeInternal(Supplier<ByteBuf> retainedBufSupplier, Callback<Long> callback) {
+    FutureResult<Long> future = new FutureResult<>();
+    ByteBuf buf = null;
+    long bytesWritten = 0;
+    Exception exception = null;
+    try {
+      if (!isOpen()) {
+        throw new ClosedChannelException();
+      } else if (totalBytesWritten.get() > sizeLimitInBytes) {
+        throw new RestServiceException("Request is larger than allowed size: " + sizeLimitInBytes,
```

Comment:
why do we throw a RestServiceException in the common package? Can we throw some more generic exception here?

Reply:
In terms of production code, this is mostly used by the rest/frontend package and throwing RestServiceException directly makes error handling cleaner (correct response code generated without try/catch).

Reply:
One other option would be to move it to ambry-rest, and get rid of `InputStreamReadableStreamChannel`, which has no usages (other than a unit test)